### PR TITLE
feat: Add in-dialog command whitelisting with LLM suggestions (#5480)

### PR DESCRIPTION
**Pull request by hannesrudolph**
image

By Default both the output of the terminal and the command permissions are zipped up. 

image

image

image

### Related GitHub Issue

Closes: #5480 

### Description

This PR enhances the command approval dialog by adding an in-context whitelisting interface that allows users to quickly add command patterns to their auto-approve list without navigating away to settings.

**Key Implementation Details:**
- Implemented deterministic command pattern extraction using the `shell-quote` library for proper command parsing
- Created a new `CommandExecution` component that renders the command approval UI with a collapsible whitelist section
- Automatically generates command pattern suggestions based on the actual command structure (base command and full command)
- Integrated with the existing `allowedCommands` settings mechanism for immediate persistence
- Added comprehensive translations for all 18 supported languages

**Design Choices:**
- Used `shell-quote` library instead of LLM suggestions for reliable, deterministic command parsing (addresses reviewer feedback)
- Automatically extracts both the base command (e.g., "npm") and full command pattern (e.g., "npm install") as suggestions
- Limited suggestions to 2-3 patterns to prevent overwhelming users
- Used a collapsible UI to keep the approval dialog clean when whitelisting isn't needed
- Filtered out already-whitelisted patterns to avoid redundancy
- Used checkboxes for explicit user control over which patterns to whitelist

**Areas for Review Focus:**
- The `commandPatterns.ts` utility that uses `shell-quote` for command parsing
- The `CommandExecution` component implementation and its integration with the settings system
- The UI/UX of the collapsible whitelist section
- Translation consistency across all languages

### Test Procedure

**Unit Tests Added:**
- `commandPatterns.spec.ts`: Tests for the deterministic command pattern extraction logic
- `CommandExecution.spec.tsx`: Component tests for rendering suggestions, handling selections, and updating settings
- `webviewMessageHandler.spec.ts`: Tests for handling the new message types

**Manual Testing Steps:**
1. Trigger a command that requires approval (e.g., ask Roo to run `npm install`)
2. Verify the command approval dialog appears with the new collapsible "Add to Allowed Auto-Execute Commands" section
3. Expand the section and verify automatically extracted patterns appear as checkboxes (e.g., "npm", "npm install")
4. Select one or more patterns and click "Add Selected Patterns"
5. Verify the patterns are added to settings (check Settings > Roo Code > Auto-approve)
6. Run the same command again and verify it's auto-approved
7. Test edge cases:
   - Complex commands with arguments and flags
   - Commands with quoted arguments
   - Commands where all patterns are already whitelisted
   - Commands with special characters

**Testing Environment:**
- VSCode version: Latest stable
- Extension running in development mode
- Multiple language settings to verify translations

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

[Please add screenshots showing:
1. The command approval dialog with the collapsed whitelist section
2. The expanded whitelist section showing automatically extracted patterns with checkboxes
3. The settings panel showing newly added patterns]

### Documentation Updates

- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).

### Additional Notes

**Translation Updates:**
This PR includes translations for all 18 supported languages. The new translation keys added are:
- `chat.command.addToWhitelist`: Label for the collapsible section
- `chat.command.whitelistDescription`: Instructional text above checkboxes
- `chat.command.addSelected`: Button label for adding selected patterns
- `common.info.command_whitelisted`: Confirmation message after whitelisting

All translations maintain the informal voice consistent with the project style.

**Implementation Note:**
Based on reviewer feedback, this PR now uses a deterministic approach with the `shell-quote` library for command parsing instead of relying on LLM suggestions. This ensures consistent and reliable pattern extraction that properly handles complex command structures, quoted arguments, and special characters.

### Get in Touch

[Your Discord username]

----

> [!IMPORTANT]
> Adds in-dialog command whitelisting with LLM suggestions, enhancing command parsing and UI for managing command permissions.
> 
>   - **Behavior**:
>     - Adds in-dialog command whitelisting interface in `CommandExecution.tsx` for quick pattern addition to auto-approve list.
>     - Uses `shell-quote` for deterministic command parsing in `commandPatterns.ts`.
>     - Integrates with `allowedCommands` settings for immediate persistence.
>   - **UI Components**:
>     - New `CommandPatternSelector` component for managing command permissions.
>     - Updates `CommandExecution.tsx` to include collapsible whitelist section.
>   - **Command Parsing**:
>     - Refactors command parsing logic into `commandUtils.ts`.
>     - Implements `extractCommandPatterns()` in `commandPatterns.ts` for pattern extraction.
>   - **Tests**:
>     - Adds tests in `commandPatterns.spec.ts` for pattern extraction logic.
>     - Updates `executeCommandTool.spec.ts` for handling suggestions.
>   - **Misc**:
>     - Removes `allowedCommands` from `package.json` configuration.
>     - Updates translations for new UI elements in multiple locales.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 83b6f3cc807d11b3320b610be7c5bae03450ef84. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/core/prompts/tools/execute-command.ts [hunk 1]
```diff
@@ -3,23 +3,66 @@ import { ToolArgs } from "./types"
 export function getExecuteCommandDescription(args: ToolArgs): string | undefined {
 	return `## execute_command
 Description: Request to execute a CLI command on the system. Use this when you need to perform system operations or run specific commands to accomplish any step in the user's task. You must tailor your command to the user's system and provide a clear explanation of what the command does. For command chaining, use the appropriate chaining syntax for the user's shell. Prefer to execute complex CLI commands over creating executable scripts, as they are more flexible and easier to run. Prefer relative commands and paths that avoid location sensitivity for terminal consistency, e.g: \`touch ./testdata/example.file\`, \`dir ./examples/model1/data/yaml\`, or \`go test ./cmd/front --config ./cmd/front/config.yml\`. If directed by the user, you may open a terminal in a different directory by using the \`cwd\` parameter.
+
+**IMPORTANT: When executing commands that match common patterns (like npm, git, ls, etc.), you SHOULD provide suggestions for whitelisting. This allows users to auto-approve similar commands in the future.**
```
Comment (KJ7LNW) [inline src/core/prompts/tools/execute-command.ts]: Use "allowlist" because "whitelist" and "blacklist" have fallen out of favor in preference of a race-inclusive lexicon.

File: src/core/prompts/tools/execute-command.ts [hunk 2]
```diff
@@ -3,23 +3,66 @@ import { ToolArgs } from "./types"
...
+- Common patterns to suggest:
+  - "npm " for all npm commands
+  - "git " for all git operations
+  - "ls " for listing files
+  - "cd " for changing directories
+  - "echo " for echo commands
+  - "mkdir " for creating directories
+  - "rm -rf node_modules" for specific cleanup command
+  - Language-specific patterns like "python ", "node ", "go test ", etc.
+  - "*" to allow all commands (only suggest when explicitly requested by user)
```
Comment (KJ7LNW) [inline src/core/prompts/tools/execute-command.ts]: Roo already has a shell-parser: I think you can implement this without system instructions at all, just parse the command and provide suggestions that, of the existing shell parser.

The idea is cute, but Roo already has the way to allow existing rules without system instruction bloat, I am sure there are users there would prefer doing this without modifying the system instructions.

If you do add a feature like this, the feature should default disabled by default so people can retain control over the size of their system instructions.

It is easy enough simply to add commands to the existing settings interface, which is fully functional and is already a zero-token feature.

If a user does enable the feature: models are pretty smart I do not think you need this much information. You and probably tell the model to "condense these requirements down to 5" lines and the result would be sufficient

File: src/core/tools/executeCommandTool.ts [hunk 3]
```diff
@@ -51,7 +51,54 @@ export async function executeCommandTool(
 			cline.consecutiveMistakeCount = 0
 
 			command = unescapeHtmlEntities(command) // Unescape HTML entities.
-			const didApprove = await askApproval("command", command)
+
+			// Parse suggestions if provided
```
[the remark this answers is not in the document]
Comment (KJ7LNW) [reply src/core/tools/executeCommandTool.ts]: I agree with this parsing should be centralized in Roo's core.

Alternatively, cherry-pick this commit which provides a generalized way of passing metadata down from the tool into the web view:

https://github.com/RooCodeInc/Roo-Code/pull/2670/commits/a2bd68b0a04712d5dd399112f9034a827f95c3fd

File: src/core/webview/webviewMessageHandler.ts [hunk 4]
```diff
@@ -776,6 +776,36 @@ export const webviewMessageHandler = async (
 
 			break
 		}
+		case "whitelistCommand": {
```
[the remark this answers is not in the document]
Comment (KJ7LNW) [reply src/core/webview/webviewMessageHandler.ts]: allowedCommands - is better, avoid the word "white" list

File: src/core/webview/webviewMessageHandler.ts [hunk 5]
```diff
@@ -776,6 +776,36 @@ export const webviewMessageHandler = async (
...
+				const currentCommands = getGlobalState("allowedCommands") ?? []
+				const validCommands = Array.isArray(currentCommands)
+					? currentCommands.filter((cmd) => typeof cmd === "string" && cmd.trim().length > 0)
+					: []
+
+				// Add the new pattern if it's not already in the list
+				if (!validCommands.includes(message.pattern)) {
+					validCommands.push(message.pattern)
+
+					await updateGlobalState("allowedCommands", validCommands)
```
Comment (KJ7LNW) [inline src/core/webview/webviewMessageHandler.ts]: `await updateGlobalState("allowedCommands", validCommands)` is used throughout the code base update the allowed list of commands, but there should be a function for this purpose instead of directly hacking the global state via `updateGlobalState`

File: webview-ui/src/utils/commandParsing.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+export const parseCommandAndOutput = (text: string | undefined): ParsedCommand => {
+	if (!text) {
+		return { command: "", output: "", suggestions: [] }
+	}
+
+	// First, extract suggestions from the text
+	const suggestions: string[] = []
+
+	// Parse <suggestions> tag with JSON array
+	const suggestionsMatch = text.match(/<suggestions>([\s\S]*?)<\/suggestions>/)
```
Comment (KJ7LNW) [inline webview-ui/src/utils/commandParsing.ts]: Parser v2 (v3 v4 v5?) should really handle all parsing considerations.

but again I do not think the model should supply these: the existing shell parser that is already used for command validation should be sufficient to find excerpts of commands to be allowed

File: webview-ui/src/utils/commandPatterns.ts [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+ * - "npm run build" -> "npm run"
+ * - "git commit -m 'message'" -> "git commit"
+ * - "echo 'hello world'" -> "echo"
+ * - "python script.py --arg value" -> "python"
+ * - "./scripts/deploy.sh production" -> "./scripts/deploy.sh"
+ * - "cd /path/to/dir && npm install" -> "cd * && npm install"
+ * - "rm -rf node_modules" -> "rm"
+ */
+export function extractCommandPattern(command: string): string {
+	if (!command?.trim()) return ""
```
Comment (KJ7LNW) [inline webview-ui/src/utils/commandPatterns.ts]: yuck! use a shell parser -- reinventing the wheel only increases afloat and poorly implements parsing.

the commands you need to support could be compound commands:

```ts
./execall git branch -a |grep hierarchical-rule-loading | while read r _ b; do (cd $r && git log $b src/core/prompts/sections/__tests__/custom-instructions.spec.ts_; done
```

File: webview-ui/src/utils/commandPatterns.ts [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+	}
+
+	// Not a chained command, process normally
+	return extractSingleCommandPattern(trimmedCommand)
+}
+
+/**
+ * Extracts pattern from a single command (not chained)
+ */
+function extractSingleCommandPattern(command: string): string {
```
Comment (KJ7LNW) [inline webview-ui/src/utils/commandPatterns.ts]: delete this entire file and use a library

**Review by KJ7LNW — COMMENTED** [review]
Review comments are below, generally speaking I think this could be improved by using a library that parses the original command into its components for suggestions; this allows us to exclude system instructions and command a tool changes altogether:

See the existing command parser here, which uses the `shell-quote` library that is only available in the code base:

- [webview-ui/src/utils/command-validation.ts](https://github.com/RooCodeInc/Roo-Code/blob/main/webview-ui/src/utils/command-validation.ts)

using the existing parser (or another command line parsing library) you can more easily extract commands that are provided by the model without involving the model at all. 

let's first try and implement this without system instructions at all. if that proves not to be possible, then please default the new system instructions to be _disabled_ because users can already allow commands manually without affecting the length of their instruction tokens.

**Comment by hannesrudolph** [comment]
> Review comments are below, generally speaking I think this could be improved by using a library that parses the original command into its components for suggestions; this allows us to exclude system instructions and command a tool changes altogether:
> 
> 
> 
> See the existing command parser here, which uses the `shell-quote` library that is only available in the code base:
> 
> 
> 
> - [webview-ui/src/utils/command-validation.ts](https://github.com/RooCodeInc/Roo-Code/blob/main/webview-ui/src/utils/command-validation.ts)
> 
> 
> 
> using the existing parser (or another command line parsing library) you can more easily extract commands that are provided by the model without involving the model at all. 
> 
> 
> 
> let's first try and implement this without system instructions at all. if that proves not to be possible, then please default the new system instructions to be _disabled_ because users can already allow commands manually without affecting the length of their instruction tokens.

I disagree. I already attempted without the LLM and it was not robust. I like this implementation. Code needs work still.

**Comment by KJ7LNW** [comment]
> I disagree. I already attempted without the LLM and it was not robust. I like this implementation. Code needs work still.

Can we leave added tokens disabled by default?

**Comment by hannesrudolph** [comment]
Roo working better by default is more imperative than lower tokens. Our base system prompt still comes in lower than the base Claude Code one and they clearly recognize that a good system prompt is important. The added tokens are needed to get the suggestion no?

File: src/core/prompts/tools/execute-command.ts (hunk 2 above)
Comment (hannesrudolph) [reply src/core/prompts/tools/execute-command.ts]: I do agree some bloat can likely be removed from this blurb, but i disagree with everything else you said.

**Comment by KJ7LNW** [comment]
> ... but i disagree with everything else you said.

If you have already made up your mind, then there is nothing I can do. 

If one of my enterprise customers came to me and asked the same question, this is the analysis that I would give:

## Zero-token Implementation

It is possible to make a robust zero-token mechanism with very little code.

### Parser
```js
import { parse } from 'shell-quote';

function extractPatterns(cmdStr) {
    const patterns = new Set();
    
    const parsed = parse(cmdStr);
    
    const commandSeparators = new Set(['|', '&&', '||', ';']);
    let current = [];
    for (const token of parsed) {
        if (typeof token === 'object' && token.op && commandSeparators.has(token.op)) {
            if (current.length) processCommand(current, patterns);
            current = [];
        } else {
            current.push(token);
        }
    }

    if (current.length) processCommand(current, patterns);
        
    return patterns;
}
```


### Filter
```js
function processCommand(cmd, patterns) {
    if (!cmd.length || typeof cmd[0] !== 'string') return;
    
    const mainCmd = cmd[0];
    patterns.add(mainCmd);
    
    const breakingExps = [ /^-/, /[\\/:.~ ]/, ];
    
    for (let i = 1; i < cmd.length; i++) {
        const arg = cmd[i];

        if (typeof arg !== 'string' || breakingExps.some(re => re.test(arg))) break;
        
        const pattern = cmd.slice(0, i + 1).join(' ');
        patterns.add(pattern);
    }
}
```

### Display
```js
const exampleCommands = [ 'git ... ' , ... ]
const allPatterns = new Set();
exampleCommands.forEach(cmd => {
    extractPatterns(cmd).forEach(pattern => allPatterns.add(pattern));
});

const sorted = Array.from(allPatterns).sort((a, b) => {
    return a.localeCompare(b);
});

console.log('All suggested command patterns for auto-allowing (sorted and unique):', sorted);
```

### Example Outputs 

```js
[
  '../repackage.sh',
  '../roo-main/build.sh',
  '../roo-main/clean.sh',
  'cd',
  'code',
  'code-insiders',
  'fg',
  'git',
  'git add',
  'git checkout',
  'git cherry-pick',
  'git diff',
  'git diff HEAD',
  'git fetch',
  'git fetch github',
  'git fetch origin',
  'git log',
  'git push',
  'git push github',
  'git push github refactor-use-files-for-history',
  'git rebase',
  'git reset',
  'git restore',
  'git show',
  'git stash',
  'git stash pop',
  'git status',
  'grep',
  'grep map',
  'head',
  'npm',
  'npm run',
  'npm run start',
  'npx',
  'npx jest',
  'pnpm',
  'pnpm test',
  'top',
  'unzip',
  'vi',
  'xclip'
]
```

### Example Inputs

The following examples are based on the top 50 commands from a developer's shell history. As you can see some of them are rather complex:

```sh
      2 unzip -l builds/roo-cline-3.21.5-error-boundary-component.vsix|grep map
      3 cd src/
      3 cd webview-ui/
      3 git add webview-ui/
      3 git checkout -b use-safe-write-json-for-all-files
      3 git checkout v3.16.5 .roomodes
      3 git cherry-pick --continue
      3 git diff HEAD
      3 git diff pnpm-lock.yaml
      3 git fetch github 
      3 git push  github refactor-use-files-for-history -f
      3 git rebase origin/main 
      3 git restore .roomodes 
      3 git restore --staged .roomodes
      3 git show f65fb69fd~1 f65fb69fdfb60abb0968b658c320ae5adb1e3d8b
      3 npx jest core/prompts/sections/__tests__/custom-instructions.test.ts
      3 top
      3 unzip -l builds/roo-cline-3.21.5-error-boundary-component.vsix|grep map|head
      4 code --install-extension builds/roo-cline-3.21.5-error-boundary-component.vsix
      4 git add webview-ui/src/App.tsx
      4 git checkout github/use-safe-write-json-for-all-files
      4 git checkout HEAD:.roomodes
      4 git diff origin/main
      4 git diff --stat
      4 git diff webview-ui/
      4 git diff|xclip
      4 git rebase origin/main
      4 git stash
      4 git stash pop
      5 cd -
      5 code
      5 git checkout .roomodes
      5 git diff origin/main --stat
      6                         }
      6 git log 
      6 git log --stat
      6 git rebase v3.23.5
      6 ../roo-main/clean.sh 
      6 vi ../repackage.sh 
      7 git diff -w
      7 ../repackage.sh 
      8 git diff 
      9 ../roo-main/build.sh 
     10 cd ..
     10 code-insiders .
     11 git checkout origin/main
     11 git fetch origin 
     11 git reset --hard
     13 ../roo-main/build.sh  && code --install-extension builds/roo-cline-3.21.5-error-boundary-component.vsix && code
     15 git status 
     17 fg
     19 pnpm test
     41 git log
     49 git diff
    114 git status
```

File: src/core/prompts/tools/execute-command.ts (hunk 2 above)
Comment (hannesrudolph) [reply src/core/prompts/tools/execute-command.ts]: The shell parser simply is not effective enough.

**Comment by hannesrudolph** [comment]
@KJ7LNW Thanks for the suggestion! The zero-token shell parser approach using `shell-quote` is an interesting optimization that would avoid the need for LLM-generated suggestions entirely. This would indeed be more efficient and could handle complex compound commands better.

For now, the current implementation addresses the immediate need and is working correctly. We'll consider refactoring to use a shell parser approach in a future iteration, which would:
- Eliminate the need for system instruction tokens
- Better handle compound commands like your example
- Provide more accurate pattern extraction

I appreciate the detailed feedback and the code reference to the existing command parser!
